### PR TITLE
sys-libs/libcxxabi: new package

### PR DESCRIPTION
**Pull request by lei-zz**
Changes introduced in this PR:
- A new package: libcxxabi, which should be functionally equivalent to libcxxrt
- libcxx is rewritten to use CMake and support libcxxabi
- llvm-libunwind is enhanced to support multilib

### CODE REVIEW

**Comment by zzlei** [comment]
A few notes:
- libcxx's source is fetched because libcxxabi needs it. Logically it should be libcxx that depends on libcxxabi, not the other way around. Unfortunately these two packages aren't decoupled well, thus this little mess.
- libcxxabi's headers are installed because libcxx will need them. I'll submit a updated ebuild which adds libcxxabi support for libcxx later.
- BSD Project is listed as a proxy-maint here, for the same reason it's listed in libcxx.

**Comment by monsieurp** [comment]
@aballier is your man.

**Comment by zzlei** [comment]
A few notes on libcxx:
- libcxxabi is now the default ABI library in place of libcxxrt. This makes sense to me because libcxx and libcxxabi are both developed by LLVM and they should work pretty well together.
- Support for libsupc++ is dropped. libsupc++ support in previous ebuilds is very hackery and is no longer worth the effort IMHO, since libcxxabi and libcxxrt are good enough.
- The test phase is dropped. It's already broken in previous ebuilds as commented. I'll investigate libcxx's test infrastructure and see if I can make it work.

**Comment by blueness** [comment]
@monsieurp @aballier you can include me in any or all of those packages because this is part of @zzlei work with clang+musl.  i'll proxy for him.  i'll wait for your ack and then can go ahead and commit.

**Comment by mgorny** [comment]
How would you feel about creating an LLVM (Gentoo) project to maintain all relevant packages and have all of us in it?

**Comment by blueness** [comment]
@mgorny  that's actually a good idea.  i'm not particularly invested in llvm so i don't want to lead yet another project, but i'll be a member.

**Comment by mgorny** [comment]
I'm going to take a look at this soonish. I don't know if I'll be able to do it all at once, so possibly I'll be committing parts of it separately.

**Comment by mgorny** [comment]
@zzlei, I feel this PR is a bit too 'all-in-one' for my taste. Could you try to move at a slower pace? ;-)

Basically, what I'd like to see first is a separate commit that converts libcxx-9999 to cmake, with current set of features. No more, no less.

Second question: don't we need to multilib gnu libunwind as well?

**Comment by mgorny** [comment]
Oh, and if @aballier doesn't mind, I'd like libcxx-9999 to use git mirrors like llvm-9999 and friends do.

**Comment by mgorny** [comment]
Also, did you upstream the build system patch? Please always include bug report / patch review links, and always upstream patches that are not strictly Gentoo-specific.

**Comment by zzlei** [comment]
> Basically, what I'd like to see first is a separate commit that converts libcxx-9999 to cmake, with current set of features. No more, no less.

No problem. But I have to cut the test feature out as it doesn't work in the standalone build. From the comment above `multilib_src_test` in older ebuilds, I guess it never worked?

**Comment by mgorny** [comment]
> No problem. But I have to cut the test feature out as it doesn't work in the standalone build. From the comment above multilib_src_test in older ebuilds, I guess it never worked?

Quite possible. Feel free to cut it out, or leave broken. Preferably test the old ebuild and new one to make sure they are ~equivalent.

**Comment by zzlei** [comment]
I also cut out support for libsupc++ because it's too tricky, and isn't of much value considering libc++abi is to be supported.

**Comment by mgorny** [comment]
Yeah, that's fine. If you were really picky, you could first kill it in autotools, then cmake, but current way should work. I will review in a few minutes.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 1]
```diff
@@ -2,13 +2,13 @@
 # Distributed under the terms of the GNU General Public License v2
 # $Id$
 
-EAPI=5
+EAPI=6
 
-ESVN_REPO_URI="http://llvm.org/svn/llvm-project/libcxx/trunk"
+EGIT_REPO_URI="https://github.com/llvm-mirror/libcxx.git"
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Add the upstream git mirror first: `http://llvm.org/git/libcxx.git`

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 2]
```diff
@@ -26,84 +26,38 @@ if [ "${PV%9999}" = "${PV}" ] ; then
...
-	LD_LIBRARY_PATH="${BUILD_DIR}/lib:${LD_LIBRARY_PATH}" \
-		CC="clang++ $(get_abi_CFLAGS) ${CXXFLAGS}" \
-		HEADER_INCLUDE="-I${BUILD_DIR}/include" \
-		SOURCE_LIB="-L${BUILD_DIR}/lib" \
-		LIBS="-lm $(usex libcxxrt -lcxxrt "")" \
-		./testit || die
-	# TODO: fix link against libsupc++
+multilib_src_configure() {
+	local libdir=$(get_libdir)
+	local cxxabi=libcxxrt
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Don't use the extra variable (yet) ;-P.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 3]
```diff
@@ -26,84 +26,38 @@ if [ "${PV%9999}" = "${PV}" ] ; then
...
+
+	local mycmakeargs=(
+		-DLLVM_CONFIG_PATH=OFF
+		-DLIBCXX_LIBDIR_SUFFIX=${libdir#lib}
+		-DLIBCXX_ENABLE_SHARED=ON
+		-DLIBCXX_ENABLE_STATIC=$(usex static-libs)
+		-DLIBCXX_CXX_ABI=${cxxabi}
+		-DLIBCXX_CXX_ABI_INCLUDE_PATHS="${EPREFIX}/usr/include/${cxxabi}"
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Do we need to explicitly specify this?

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 4]
```diff
@@ -122,49 +76,42 @@ END_LDSCRIPT
...
+	# Generate libc++.a ldscript for inclusion of its dependencies so that
+	# clang++ -stdlib=libc++ -static works out of the box.
+	local deps="${EPREFIX}/usr/${libdir}/libc++_static.a ${EPREFIX}/usr/${libdir}/${cxxabi}.a"
+	# On Linux/glibc it does not link without libpthread or libdl. It is
+	# fine on FreeBSD.
+	use elibc_glibc && deps="${deps} ${EPREFIX}/usr/${libdir}/libpthread.a ${EPREFIX}/usr/${libdir}/libdl.a"
+	# unlike libgcc_s, libunwind is not implicitly linked
+	use libunwind && deps="${deps} ${EPREFIX}/usr/${libdir}/libunwind.a"
+
+	gen_ldscript "${deps}" > "${ED}/usr/${libdir}/libc++.a"
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: || die

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 5]
```diff
@@ -122,49 +76,42 @@ END_LDSCRIPT
...
+
+	# Move it first.
+	mv "${ED}/usr/${libdir}/libc++.a" "${ED}/usr/${libdir}/libc++_static.a" || die
+
+	# Generate libc++.a ldscript for inclusion of its dependencies so that
+	# clang++ -stdlib=libc++ -static works out of the box.
+	local deps="${EPREFIX}/usr/${libdir}/libc++_static.a ${EPREFIX}/usr/${libdir}/${cxxabi}.a"
+	# On Linux/glibc it does not link without libpthread or libdl. It is
+	# fine on FreeBSD.
+	use elibc_glibc && deps="${deps} ${EPREFIX}/usr/${libdir}/libpthread.a ${EPREFIX}/usr/${libdir}/libdl.a"
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: +=

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 6]
```diff
@@ -122,49 +76,42 @@ END_LDSCRIPT
...
-		dolib.a libc++.a
-		gen_static_ldscript
-	fi
-	dolib.so libc++.so*
+	cmake-utils_src_install
 	gen_shared_ldscript
+	use static-libs && gen_static_ldscript
 }
 
 multilib_src_install_all() {
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: I think it's redundant now.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 7]
```diff
@@ -122,49 +76,42 @@ END_LDSCRIPT
...
 
 multilib_src_install() {
-	cd "${BUILD_DIR}/lib"
-	if use static-libs ; then
-		dolib.a libc++.a
-		gen_static_ldscript
-	fi
-	dolib.so libc++.so*
+	cmake-utils_src_install
 	gen_shared_ldscript
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Do we still need to do that? Is there a chance to get the underlying problem solved upstream?

File: sys-libs/libcxx/libcxx-9999.ebuild (hunk 3 above)
Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Yes, and this is exactly the ugly part of supporting libsupc++. On my system, it should be set
`/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5;/usr/lib/gcc/x86_64-pc-linux-gnu/5.4.0/include/g++-v5/x86_64-pc-linux-gnu`

It's not trivial to get right and not worth the effort IMO.

File: sys-libs/libcxx/libcxx-9999.ebuild (hunk 7 above)
Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Upstream does have mechanism for this, but it has a few problems:
- like test cases, it doesn't work in a standalone build 
- it only supports generating script for shared lib, not for static lib
- it doesn't take libunwind into consideration

Some heavy patching might help, but I didn't try.

**Comment by zzlei** [comment]
I'm hitting a `Header files have changed between ABIs` error with nongnu's libunwind while adding multilib support. I'm not really interested in fixing its headers...

**Comment by mgorny** [comment]
Read on `MULTILIB_WRAPPED_HEADERS` in multilib-build.eclass

**Comment by zzlei** [comment]
The _nongnu_ libunwind is now updated with multilib support.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 8]
```diff
@@ -26,84 +27,37 @@ if [ "${PV%9999}" = "${PV}" ] ; then
...
-}
-
-multilib_src_compile() {
-	cd "${BUILD_DIR}/lib" || die
-	emake shared
-	use static-libs && emake static
-}
-
-# Tests fail for now, if anybody is able to fix them, help is very welcome.
-multilib_src_test() {
```
Comment (aballier) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Tests should be kept; fixed if possible, but definitely not dropped.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 9]
```diff
@@ -26,84 +27,37 @@ if [ "${PV%9999}" = "${PV}" ] ; then
 else
 	KEYWORDS=""
 fi
-IUSE="elibc_glibc elibc_musl +libcxxrt libunwind +static-libs test"
-REQUIRED_USE="libunwind? ( libcxxrt )"
+IUSE="elibc_glibc elibc_musl libunwind +static-libs"
```
Comment (aballier) [inline sys-libs/libcxx/libcxx-9999.ebuild]: is there any reason to drop support for libsupc++ ?

Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: I think the location of its headers is too tricky to find. And given the options of libcxxrt and libc++abi, are there many people who would use libc++ with libsupc++? If so, I could definitely bring it back.

Comment (aballier) [inline sys-libs/libcxx/libcxx-9999.ebuild]: yeah but so far it works and upstream supports it so i see no reason to drop it

File: sys-libs/libcxx/libcxx-9999.ebuild (hunk 8 above)
Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Doesn't it confuse users when tests fail for no good reason? In fact a user once reported this failure to me and he thought the compiler is broken.

IMHO it's not too late to bring them back when they actually work.

**Comment by zzlei** [comment]
Updates:
- support for libsupc++ is brought back; the mechanism for locating its headers is simpler than the original version, but it requires `CC=gcc` to work correctly
- to be more conservative, libcxxrt is kept as the default C++ ABI library

File: sys-libs/libcxx/libcxx-9999.ebuild (hunk 8 above)
Comment (aballier) [inline sys-libs/libcxx/libcxx-9999.ebuild]: it is better to have failing tests than no tests at all (<=> tests pass) when they actually fail

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 10]
```diff
@@ -26,84 +27,64 @@ if [ "${PV%9999}" = "${PV}" ] ; then
...
-			[ -n "${found}" ] || die "Header not found: ${i}"
-		done
+		local gcc_inc="${EPREFIX}/usr/lib/gcc/${CHOST}/$(gcc-fullversion)/include/g++-v$(gcc-major-version)"
+		cxxabi=libsupc++
+		cxxabi_incs="${gcc_inc};${gcc_inc}/${CHOST}"
 	fi
 
-	tc-export AR CC CXX
-
-	append-ldflags "-Wl,-z,defs" # make sure we are not underlinked
```
Comment (aballier) [inline sys-libs/libcxx/libcxx-9999.ebuild]: is this default in cmake ?

Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: It seems not. Nevertheless I'm using ld.gold and don't see any underlinking. Is it really a problem?

Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: If upstream libcxx doesn't use it, submit a patch. AFAICS llvm and clang are already doing this, so it's just a matter of copying the appropriate line from cmake files.

Comment (zzlei) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Patch submitted: https://reviews.llvm.org/D24119

Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Err, you were supposed to put -commits as subscriber, and guess reviewers from `git blame`.

**Comment by mgorny** [comment]
Ok, I'm done with compiler-rt, so I'll get to those packages soon.

**Comment by zzlei** [comment]
Update:
- install llvm-libunwind's headers into /usr/include/libunwind
- fix libc++abi to actually use llvm-libunwind's unwind.h, instead of clang/gcc's

**Comment by zzlei** [comment]
A side note: during the configure phase, both libc++ and libc++abi emit a **harmless** warning. Should I fix their cmake scripts to suppress the warning?

**Comment by zzlei** [comment]
I've submitted libunwind related changes in a new PR: https://github.com/gentoo/gentoo/pull/2212

This PR remains merely for discussion. Please don't merge it.

**Comment by mgorny** [comment]
Ok, since we have libcxx in-tree finally, I'm going to take a look at this. @lei-zz, no need to do anything on your end yet ;-).

**Comment by mgorny** [comment]
I'm going to close this PR not to overwrite your commits and start over.

**Pull request closed without merging**